pkg/service/containermanager/docker.go

```diff
@@ -0,0 +1,210 @@
+	instanceScope *scope.InstanceScope
+
+	templateFactory func(sshClient ssh.Interface, template *template.Template, data file.Data, dst string) (operation.Template, error)
+	dockerFactory   func(sshClient ssh.Interface, version, arch string) (operation.Binary, error)
+}
+
+func NewDockerService(sshClient ssh.Interface, scope scope.KKInstanceScope, instanceScope *scope.InstanceScope) *DockerService {
+	return &DockerService{
+		sshClient:     sshClient,
+		scope:         scope,
+		instanceScope: instanceScope,
+	}
+}
+
+func (d *DockerService) getTemplateService(template *template.Template, data file.Data, dst string) (operation.Template, error) {
+	if d.templateFactory != nil {
+		return d.templateFactory(d.sshClient, template, data, dst)
+	}
+	return file.NewTemplate(d.sshClient, d.scope.RootFs(), template, data, dst)
+}
+
+func (d *DockerService) getDockerService(sshClient ssh.Interface, version, arch string) (operation.Binary, error) {
+	if d.dockerFactory != nil {
+		return d.dockerFactory(sshClient, version, arch)
+	}
+	return file.NewDocker(sshClient, d.scope.RootFs(), version, arch)
+}
+
+// Type returns the type docker of the container manager.
+func (d *DockerService) Type() string {
+	return file.DockerID
+}
+
+// Version returns the version of the container manager.
+func (d *DockerService) Version() string {
+	return d.instanceScope.KKInstance.Spec.ContainerManager.Version
+}
+
+// IsExist returns true if the container manager is installed.
+func (d *DockerService) IsExist() bool {
+	res, err := d.sshClient.SudoCmd(
+		"if [ -z $(which docker) ] || [ ! -e /run/docker.sock ]; " +
+			"then echo 'not exist'; " +
+			"fi")
+	if err != nil {
+		return false
+	}
+	if strings.Contains(res, "not exist") {
+		return false
+	}
+	return true
+}
+
+func (d *DockerService) Get(timeout time.Duration) error {
+	docker, err := d.getDockerService(d.sshClient, d.Version(), d.instanceScope.Arch())
+	if err != nil {
+		return err
+	}
+	binaries := []operation.Binary{
+		docker,
+	}
+	zone := d.scope.ComponentZone()
+	host := d.scope.ComponentHost()
+	overrideMap := make(map[string]infrav1.Override)
+	for _, o := range d.scope.ComponentOverrides() {
+		overrideMap[o.ID+o.Version+o.Arch] = o
+	}
+
+	for _, b := range binaries {
+		d.instanceScope.V(4).Info("download binary", "binary", b.Name(), "version", b.Version(),
+			"url", b.URL().String())
+
+		override := overrideMap[b.ID()+b.Version()+b.Arch()]
+		if err := util.DownloadAndCopy(b, zone, host, override.Path, override.URL, override.Checksum.Value, timeout); err != nil {
+			return err
+		}
+	}
+
+	// /usr/local
+	dir := filepath.Dir(filepath.Dir(docker.RemotePath()))
+	if _, err := d.sshClient.SudoCmdf("tar Cxzvf %s %s && mv %s/docker/* %s", dir, docker.RemotePath(), dir, directory.BinDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *DockerService) Install() error {
+	if err := d.installDocker(); err != nil {
+		return err
+	}
+	// TODO install https://github.com/Mirantis/cri-dockerd
```

Comment:
@24sama Do we need to install [cri-dockerd](https://github.com/Mirantis/cri-dockerd) ?

Reply:
> I think capkk can determine if need to install cri-dockerd based on the Kubernetes version

Yes, we can judge the kuberentes version

> Do you think we need to add some fields in the [api](https://github.com/kubesphere/kubekey/blob/master/api/v1beta1/containermanager_types.go) or just hard-code its version?

Yes, and also need a chinese download link